azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -59,95 +62,166 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
-
-        self.incarnation = findtext(xml_doc, "Incarnation")
-        self.expected_state = findtext(xml_doc, "ExpectedState")
-        role_instance = find(xml_doc, "RoleInstance")
+        self._xml_doc = parse_doc(self.xml_text)
+
+        self._ext_config_retriever = ext_config_retriever
+        self._wire_client = wire_client
+
+        # Cached properties
+        # Implementation note: cached_property was not used because it considers a None value to mean the property
+        # wasn't retrieved
+        self._ext_conf = None
+        self._ext_conf_retrieved = False
+        self._hosting_env = None
+        self._hosting_env_retrieved = False
+        self._shared_conf = None
+        self._shared_conf_retrieved = False
+        self.certs = None
+        self._remote_access = None
+        self._remote_access_retrieved = False
+        self._artifacts_profile_blob_url = None
+        self._status_upload_blob_url = None
+        self._status_upload_blob_type = None
+        self._ext_conf_properties_retrieved = False
+
+        self.incarnation = findtext(self._xml_doc, "Incarnation")
+        self.expected_state = findtext(self._xml_doc, "ExpectedState")
+        role_instance = find(self._xml_doc, "RoleInstance")
         self.role_instance_id = findtext(role_instance, "InstanceId")
         role_config = find(role_instance, "Configuration")
         self.role_config_name = findtext(role_config, "ConfigName")
-        container = find(xml_doc, "Container")
+        container = find(self._xml_doc, "Container")
         self.container_id = findtext(container, "ContainerId")
-        lbprobe_ports = find(xml_doc, "LBProbePorts")
+        lbprobe_ports = find(self._xml_doc, "LBProbePorts")
         self.load_balancer_probe_port = findtext(lbprobe_ports, "Port")
 
-        GoalState.ContainerID = self.container_id
-
-        fetch_full_goal_state = False
-        if full_goal_state:
-            fetch_full_goal_state = True
-            reason = 'force update'
-        elif base_incarnation is not None and self.incarnation != base_incarnation:
-            fetch_full_goal_state = True
-            reason = 'new incarnation'
-
-        if not fetch_full_goal_state:
-            self.hosting_env = None
-            self.shared_conf = None
-            self.certs = None
-            self.ext_conf = None
-            self.remote_access = None
-            return
-
-        logger.info('Fetching new goal state [incarnation {0} ({1})]', self.incarnation, reason)
-
-        try:
-            uri = findtext(xml_doc, "HostingEnvironmentConfig")
-            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-            self.hosting_env = HostingEnv(xml_text)
+        # Nothing retrieves certificates, so we need to keep this logic the same for now
+        # Limit retrieving certificates to only once per incarnation
+        uri = findtext(self._xml_doc, "Certificates")
+        if uri is None:
+            GoalState._Certs = None
```

Comment:
I dont understand the use of extra static variables to figure out if incarnation changed? Cant you just use the same logic as below to figure out if Incarnation changed and download certs if they did? 

i.e. 
`current_goal_state is not None and self.incarnation != current_goal_state.incarnation:`

Reply:
Just tried this and was reminded why in a unit test. :)
This hits a specific unit test and required quite a bit of tinkering to get right.

Reply:
+1 for use of statics opening the code to race conditions